### PR TITLE
Add guidelines about implementing interfaces.

### PR DESCRIPTION
**Pull request by munificent**
As with superclasses and mixins, you shouldn't implement a class's
interface unless the class says it's OK.

Also a bit of other copy-editing.

Fix #139.

### CODE REVIEW

**Comment by kwalrath** [comment]
Looking now...

File: src/_guides/language/effective-dart/design.md [hunk 1]
```diff
@@ -581,13 +582,41 @@ doesn't do that, it's best to assume you should *not* extend the class.
...
+usually a safe, non-breaking change. But if you are implementing that class's
+interface, now your class has a static error because it lacks an implementation
+of that new method.
+
+Package maintainers need the ability to evolve existing classes without breaking
+users. Treating *every* class as exposing an interface you are free to implement
+makes that very difficult for them. That difficulty in turn means the packages
+you rely on are slower to grow and adapt to new needs.
+
+To give the authors of the classes you use more leeway, avoid implementing their
```
Comment (kwalrath) [inline src/_guides/language/effective-dart/design.md]: Unclear what "their" refers to. How about:

To give the authors of the classes you use more leeway, avoid implementing implicit interfaces except for classes that are clearly intended to be implemented.

File: src/_guides/language/effective-dart/design.md [hunk 2]
```diff
@@ -581,13 +582,41 @@ doesn't do that, it's best to assume you should *not* extend the class.
...
+implicit interface unless the class is clearly intended to be used in that way.
+Otherwise, you may introduce a coupling that the author doesn't intend and they
```
Comment (kwalrath) [inline src/_guides/language/effective-dart/design.md]: and -> , and

File: src/_guides/language/effective-dart/design.md [hunk 3]
```diff
@@ -581,13 +582,41 @@ doesn't do that, it's best to assume you should *not* extend the class.
...
+may break your code without realizing it.
+
+### DO document if your class supports being used as an interface.
+
+Mention in the class's doc comment if the class can be used as an interface.
```
Comment (kwalrath) [inline src/_guides/language/effective-dart/design.md]: I found this a little hard to read. Maybe switch this around:

If your class can be used as an interface, mention that in the class's doc comment.

File: src/_guides/language/effective-dart/design.md [hunk 4]
```diff
@@ -470,17 +470,18 @@ commitment on your library's part to support that and behave properly when it
...
 doesn't know if any code outside of its view is using it.
 
 
-## Types
+## Classes
 
-Dart supports a variety of built-in types and you can define your own types.
-Or, you can choose not to use types at all.
+Dart is a "pure" object-oriented language in that all objects are instances of
+classes. But it does not require all code to defined inside a class -- you can
```
[the remark this answers is not in the document]
Comment (munificent) [reply src/_guides/language/effective-dart/design.md]: Done. Just changed "it" -> "Dart".

File: src/_guides/language/effective-dart/design.md [hunk 5]
```diff
@@ -581,13 +582,41 @@ doesn't do that, it's best to assume you should *not* extend the class.
...
+But implementing a class's interface is a very tight coupling to that class. It
+means virtually *any* change to the class whose interface you are implementing
+will break your implementation. For example, adding a new member to a class is
+usually a safe, non-breaking change. But if you are implementing that class's
+interface, now your class has a static error because it lacks an implementation
+of that new method.
+
+Package maintainers need the ability to evolve existing classes without breaking
+users. Treating *every* class as exposing an interface you are free to implement
+makes that very difficult for them. That difficulty in turn means the packages
```
[the remark this answers is not in the document]
Comment (munificent) [reply src/_guides/language/effective-dart/design.md]: Done.

File: src/_guides/language/effective-dart/design.md [hunk 6]
```diff
@@ -581,13 +582,41 @@ doesn't do that, it's best to assume you should *not* extend the class.
...
+implementation of that contract.
+
+But implementing a class's interface is a very tight coupling to that class. It
+means virtually *any* change to the class whose interface you are implementing
+will break your implementation. For example, adding a new member to a class is
+usually a safe, non-breaking change. But if you are implementing that class's
+interface, now your class has a static error because it lacks an implementation
+of that new method.
+
+Package maintainers need the ability to evolve existing classes without breaking
```
[the remark this answers is not in the document]
Comment (munificent) [reply src/_guides/language/effective-dart/design.md]: Done.

File: src/_guides/language/effective-dart/design.md (hunk 1 above)
Comment (munificent) [reply src/_guides/language/effective-dart/design.md]: Done.

File: src/_guides/language/effective-dart/design.md (hunk 2 above)
Comment (munificent) [reply src/_guides/language/effective-dart/design.md]: Done.

File: src/_guides/language/effective-dart/design.md (hunk 3 above)
Comment (munificent) [reply src/_guides/language/effective-dart/design.md]: Done.

**Comment by munificent** [comment]
Thanks!

**Pull request merged**